lib/trie/trie.go

```diff
@@ -167,675 +177,895 @@ func (t *Trie) Entries() map[string][]byte {
-	switch c := curr.(type) {
-	case *node.Branch:
-		fullKey := append(prefix, c.Key...)
-		var cmp int
-		if len(key) < len(fullKey) {
-			if bytes.Compare(key, fullKey[:len(key)]) == 1 { // arg key is greater than full, return nil
-				return nil
-			}
+func findNextKey(parent Node, prefix, searchKey []byte) (nextKey []byte) {
+	if parent == nil {
+		return nil
+	}
 
-			// the key is lexicographically less than the current node key. return first key available
-			cmp = 1
-		} else {
-			// if cmp == 1, then node key is lexicographically greater than the key arg
-			cmp = bytes.Compare(fullKey, key[:len(fullKey)])
-		}
+	if parent.Type() == node.LeafType {
+		parentLeaf := parent.(*node.Leaf)
+		return findNextKeyLeaf(parentLeaf, prefix, searchKey)
+	}
 
-		// if length of key arg is less than branch key,
-		// return key of first child, or key of this branch,
-		// if it's a branch with value.
-		if (cmp == 0 && len(key) == len(fullKey)) || cmp == 1 {
-			if c.Value != nil && bytes.Compare(fullKey, key) > 0 {
-				return fullKey
-			}
+	// Branch
+	parentBranch := parent.(*node.Branch)
+	return findNextKeyBranch(parentBranch, prefix, searchKey)
+}
 
-			for i, child := range c.Children {
-				if child == nil {
-					continue
-				}
+func findNextKeyLeaf(leaf *node.Leaf, prefix, searchKey []byte) (nextKey []byte) {
+	parentLeafKey := leaf.Key
+	fullKey := make([]byte, 0, len(prefix)+len(parentLeafKey))
+	fullKey = append(fullKey, prefix...)
+	fullKey = append(fullKey, parentLeafKey...)
+
+	searchKeyBigger :=
+		(len(searchKey) < len(fullKey) &&
+			bytes.Compare(searchKey, fullKey[:len(searchKey)]) == 1) ||
+			(len(searchKey) >= len(fullKey) &&
+				bytes.Compare(searchKey[:len(fullKey)], fullKey) != -1)
+	if searchKeyBigger {
+		return nil
+	}
 
-				next := nextKey(child, append(fullKey, byte(i)), key)
-				if len(next) != 0 {
-					return next
-				}
-			}
-		}
+	return fullKey
+}
 
-		// node key isn't greater than the arg key, continue to iterate
-		if cmp < 1 && len(key) > len(fullKey) {
-			idx := key[len(fullKey)]
-			for i, child := range c.Children[idx:] {
-				if child == nil {
-					continue
-				}
-
-				next := nextKey(child, append(fullKey, byte(i)+idx), key)
-				if len(next) != 0 {
-					return next
-				}
-			}
+func findNextKeyBranch(parentBranch *node.Branch, prefix, searchKey []byte) (nextKey []byte) {
+	fullKey := make([]byte, 0, len(prefix)+len(parentBranch.Key))
+	fullKey = append(fullKey, prefix...)
+	fullKey = append(fullKey, parentBranch.Key...)
+
+	if bytes.Equal(searchKey, fullKey) {
+		const startChildIndex = 0
+		return findNextKeyChild(parentBranch.Children, startChildIndex, fullKey, searchKey)
+	}
+
+	searchKeyShorter := len(searchKey) < len(fullKey)
+	searchKeyLonger := len(searchKey) > len(fullKey)
+
+	searchKeyBigger :=
+		(searchKeyShorter &&
+			bytes.Compare(searchKey, fullKey[:len(searchKey)]) == 1) ||
+			(!searchKeyShorter &&
+				bytes.Compare(searchKey[:len(fullKey)], fullKey) != -1)
+
```

Comment:
same here lol

Reply:
Totally, done in d8b98631c716c908c8a43a294b98dae20cbf05d7